### ISSUE TITLE
Players are not in sync during playback

### ISSUE THREAD

**Issue by pratikbhiyani** [superpoweredSDK/Low-Latency-Android-Audio-iOS-Audio-Engine#353]
Hello,

I am working on an cross fade audio app. In this I am playing 7 tracks together and there is 1 second of crossfade when user select different section of tracks so there are 2 players for 1 track.

But here is issue is it go out of sync after few section switching. and sometimes there are some glitch sound as well.

**Here is my code for crossfade.**

```
// This is where the Superpowered magic happens.
static bool audioProcessing(void *clientdata, float **buffers, unsigned int inputChannels, unsigned int outputChannels, unsigned int numberOfSamples, unsigned int samplerate, uint64_t hostTime) {
    __unsafe_unretained Superpowered *self = (__bridge Superpowered *)clientdata;
    if (samplerate != self->lastSamplerate) { // Has samplerate changed?
        
        self->lastSamplerate = samplerate;
        for(int i = 0 ; i < self->arrTrack.count ; i++)
        {
            SuperpoweredAdvancedAudioPlayer *playerA = self->playersA[i];
            playerA->setSamplerate(samplerate);

             SuperpoweredAdvancedAudioPlayer *playerB = self->playersB[i];
             playerB->setSamplerate(samplerate);
        }
     };
    
    bool silence = true;
     
    for(int i = 0 ; i < self->arrTrack.count ; i++)
    {

            SuperpoweredAdvancedAudioPlayer *playerA = self->playersA[i];
            
            SuperpoweredAdvancedAudioPlayer *playerB = self->playersB[i];
            
            
            bool buffertoAdd = false;
            
            if(i>0)
                buffertoAdd = true;
            
            double msElapsedSinceLastBeatA = -1;
            
            if(i > 0)
                msElapsedSinceLastBeatA = self->playersA[i-1]->msElapsedSinceLastBeat; // When playerB needs it, playerA has already stepped this value, so save it now.
            else
                msElapsedSinceLastBeatA = self->playersA[self->arrTrack.count-1]->msElapsedSinceLastBeat;
            
            double msElapsedSinceLastBeatB = -1;
            
            if(i > 0)
                msElapsedSinceLastBeatB = self->playersB[i-1]->msElapsedSinceLastBeat; // When playerB needs it, playerA has already stepped this value, so save it now.
            else
                msElapsedSinceLastBeatB = self->playersB[self->arrTrack.count-1]->msElapsedSinceLastBeat;

            if(self -> turnPlayerA[i])
            {
             if(playerA->positionMs < (self->switchTimer[i] - ValueSubtract))
             {
                 silence = !playerA->process(self->stereoBuffer, buffertoAdd, numberOfSamples, 1.0*self->volume[i]*self->_fadeVolume*self->mute[i], playerA->currentBpm,msElapsedSinceLastBeatA);
             }
             else
             {
                 if(self->_isEnd)
                 {
                     silence = !playerB->process(self->stereoBuffer, buffertoAdd, numberOfSamples, 1.0*self->volume[i]*self->_fadeVolume*self->mute[i], playerB->currentBpm,msElapsedSinceLastBeatB);
                     
                     self -> turnPlayerA[i] =  false;
                     self->fadeCount[i] = 0;
                     
                     if(!self->_isLoop)
                         self->switchTimer[i] = playerA->durationMs;
                     else
                     {
                         if(i == 0)
                             self->_repeatToogle = !self->_repeatToogle;
                         
                         self->playersA[i]->setPosition(self->_startTimer - ValueSubtract, false, true);
                     }
                     
                     if(self->_isRepeat && (i == 0))
                         self->_repeatToogle = !self->_repeatToogle;

                     self->_isEnd = false;
                 }
                 else{
                 
                 self->fadeCount[i] = self->fadeCount[i]  + 1.0/(ValueConstant);
                 
                 if(self->fadeCount[i] > 1)
                     self->fadeCount[i] = 1;
                 
                 if (self->fadeCount[i] < 0.01f) {
                     self->volA[i] = 1.0f * headroom;
                     self->volB[i] = 0.0f;
                 } else if (self->fadeCount[i] > 0.99f) {
                     self->volA[i] = 0.0f;
                     self->volB[i] = 1.0f * headroom;
                 } else { // constant power curve
                     self->volA[i] = cosf(M_PI_2 * self->fadeCount[i]) * headroom;
                     self->volB[i] = cosf(M_PI_2 * (1.0f - self->fadeCount[i])) * headroom;
                 };
                 
                 silence = !playerA->process(self->stereoBuffer, buffertoAdd, numberOfSamples, self->volA[i]*self->volume[i]*self->_fadeVolume*self->mute[i], playerA->currentBpm,msElapsedSinceLastBeatA);
                 playerB->process(self->stereoBuffer, buffertoAdd, numberOfSamples, self->volB[i]*self->volume[i]*self->_fadeVolume*self->mute[i], playerB->currentBpm,msElapsedSinceLastBeatB);
                 
                 NSLog(@"%f",playerB->positionMs);
                
                 //if(self->fadeCount[i] == 1)
                 if(playerA->positionMs >= (self->switchTimer[i]))
                 {
                     self -> turnPlayerA[i] =  false;
                     self->fadeCount[i] = 0;
                     
                     if(!self->_isLoop)
                         self->switchTimer[i] = playerA->durationMs;
                     else
                     {
                         if(i == 0)
                             self->_repeatToogle = !self->_repeatToogle;
                         
                         self->playersA[i]->setPosition(self->_startTimer - ValueSubtract, false, true);
                     }
                     
                     if(self->_isRepeat && (i == 0))
                         self->_repeatToogle = !self->_repeatToogle;

                 }
            }
             }
            }
            else{
                if(playerB->positionMs < (self->switchTimer[i] - ValueSubtract))
                {
                    silence = !playerB->process(self->stereoBuffer, buffertoAdd, numberOfSamples, 1.0*self->volume[i]*self->_fadeVolume*self->mute[i], playerB->currentBpm,msElapsedSinceLastBeatB);
                }
                else{
                    
                    if(self->_isEnd)
                    {
                        silence = !playerA->process(self->stereoBuffer, buffertoAdd, numberOfSamples, 1.0*self->volume[i]*self->_fadeVolume*self->mute[i], playerA->currentBpm,msElapsedSinceLastBeatA);
                        
                        self -> turnPlayerA[i] =  true;
                        self->fadeCount[i] = 0;
                        
                        if(!self->_isLoop)
                            self->switchTimer[i] = playerB->durationMs;
                        else
                        {
                            if(i == 0)
                                self->_repeatToogle = !self->_repeatToogle;
                            
                            self->playersB[i]->setPosition(self->_startTimer - ValueSubtract, false, true);
                        }
                        
                        if(self->_isRepeat && (i == 0))
                            self->_repeatToogle = !self->_repeatToogle;
                        
                        self->_isEnd = false;
                    }
                    else{


                    self->fadeCount[i] = self->fadeCount[i]  + 1.0/(ValueConstant);
                    
                    if(self->fadeCount[i] > 1)
                        self->fadeCount[i] = 1;
                    
                    if (self->fadeCount[i] < 0.01f) {
                        self->volB[i] = 1.0f * headroom;
                        self->volA[i] = 0.0f;
                    } else if (self->fadeCount[i] > 0.99f) {
                        self->volB[i] = 0.0f;
                        self->volA[i] = 1.0f * headroom;
                    } else { // constant power curve
                        self->volB[i] = cosf(M_PI_2 * self->fadeCount[i]) * headroom;
                        self->volA[i] = cosf(M_PI_2 * (1.0f - self->fadeCount[i])) * headroom;
                    };
                    
                    silence = !playerA->process(self->stereoBuffer, buffertoAdd, numberOfSamples, self->volA[i]*self->volume[i]*self->_fadeVolume*self->mute[i], playerA->currentBpm,msElapsedSinceLastBeatA);
                    playerB->process(self->stereoBuffer, buffertoAdd, numberOfSamples, self->volB[i]*self->volume[i]*self->_fadeVolume*self->mute[i], playerB->currentBpm,msElapsedSinceLastBeatB);

                    NSLog(@"%f",playerA->positionMs);

//                    if(self->fadeCount[i] == 1)
                    if(playerB->positionMs >= (self->switchTimer[i]))
                    {
                        self -> turnPlayerA[i] =  true;
                        self->fadeCount[i] = 0;
                        
                        if(!self->_isLoop)
                            self->switchTimer[i] = playerB->durationMs;
                        else
                        {
                            if(i == 0)
                            self->_repeatToogle = !self->_repeatToogle;
                            
                            self->playersB[i]->setPosition(self->_startTimer - ValueSubtract, false, true);
                        }
                        
                        if(self->_isRepeat && (i == 0))
                            self->_repeatToogle = !self->_repeatToogle;
                        
                    }

                    }
                }
            }
        
    }
    
    if(self->isHalfSilence)
        silence = true;
    
    if (!silence) SuperpoweredDeInterleave(self->stereoBuffer, buffers[0], buffers[1], numberOfSamples); // The stereoBuffer is ready now, let's put the finished audio into the requested buffers.
    return !silence;
}
```

Please help and let me know if you need any clarification in code. Thank you.

**Comment by superpoweredSDK** [comment]
Always call every player instance's process() method as advised in the documentation, otherwise their internal message queue will not work. We're not able to review such long pieces of code for free, sorry.

**Comment by pratikbhiyani** [comment]
Well friend, I already tried that method before.

Before this I created different process method for every players but even that time sync issue was more so that's why I added all player in one `process()` method and thought `msElapsedSinceLastBeat` variable will help to maintain sync and it is better than before but after sometime it go out of sync. 

Do Superpowered sdk has any limit of players to play together?

It will be great if you help or suggest something.

Thanks.
